pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,289 @@
+}
+
+func (c *machineDeletionHooksController) sync(ctx context.Context, _ factory.SyncContext) error {
+	// stop if the machine API is not functional
+	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	if err := c.addDeletionHookToMasterMachines(ctx); err != nil {
+		return err
+	}
+	// attempt to remove hooks from machines pending deletion and without running etcd member
+	return c.attemptToDeleteMachineDeletionHook(ctx)
+}
+
+// it will only remove the hook once it has identified that the Machine resource is being deleted and a replacement member has been created.
+func (c *machineDeletionHooksController) attemptToDeleteMachineDeletionHook(ctx context.Context) error {
+	desiredControlPlaneReplicasCount, err := readDesiredControlPlaneReplicasCount(c.configMapListerForKubeSystemNamespace)
+	if err != nil {
+		return err
+	}
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	// machines with the deletion hooks are the ones that host etcd members
+	memberMachines := filterMachinesWithMachineDeletionHook(masterMachines)
+
+	// if the number of master machines is > controlPlane.replicas based on data from the cache
+	if len(memberMachines) <= desiredControlPlaneReplicasCount {
+		klog.V(4).Infof("haven't found a replacement machine, the number of desired control plane replicas: %d must be greater than the current number of machines that host an etcd member: %d", desiredControlPlaneReplicasCount, len(memberMachines))
+		return nil
+	}
+
+	// if at least one member machine is pending deletion
+	memberMachinesPendingDeletion := filterMachinesPendingDeletion(memberMachines)
+	if len(memberMachinesPendingDeletion) == 0 {
+		klog.V(4).Infof("currently, we don't have a machine pending deletion")
+		return nil
+	}
+
+	// get the members, issue a live call instead of reading from the config map
+	// because we need to take into account learners
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return err
+	}
+	currentMemberIPListSet := sets.NewString()
+	for _, member := range members {
+		votingMemberIP, err := ceohelpers.MemberToNodeInternalIP(member)
+		if err != nil {
+			return err
+		}
+		currentMemberIPListSet.Insert(votingMemberIP)
+	}
+
+	// find a machine that doesn't have a member
+	for memberMachinePendingDeletionIP, memberMachinePendingDeletion := range indexMachinesByNodeInternalIP(memberMachinesPendingDeletion) {
+		if !currentMemberIPListSet.Has(memberMachinePendingDeletionIP) {
+			memberMachinePendingDeletionCopy := memberMachinePendingDeletion.DeepCopy()
+			removeMachineDeletionHook(memberMachinePendingDeletionCopy)
+			if _, err := c.machineClient.Update(ctx, memberMachinePendingDeletionCopy, metav1.UpdateOptions{}); err == nil {
+				klog.Infof("successfully removed the deletion hook from machine %v", memberMachinePendingDeletion.Name)
+			}
+			return err
+		}
```

Comment:
Style nit: You don't have to change it :) but I find these blocks more readable when we can return/continue early to reduce nesting if possible.
```suggestion
		if currentMemberIPListSet.Has(memberMachinePendingDeletionIP) {
			// Ignore machines that still have a member present
			continue
		}
		memberMachinePendingDeletionCopy := memberMachinePendingDeletion.DeepCopy()
		removeMachineDeletionHook(memberMachinePendingDeletionCopy)
		if _, err := c.machineClient.Update(ctx, memberMachinePendingDeletionCopy, metav1.UpdateOptions{}); err == nil {
			klog.Infof("successfully removed the deletion hook from machine %v", memberMachinePendingDeletion.Name)
		}
		return err
	}
```

Reply:
np